esp/smc/SMCLib/TpWrapper.cpp

```diff
@@ -2213,3 +2213,68 @@ extern TPWRAPPER_API void initContainerRoxieTargets(MapStringToMyClass<ISmartSoc
     }
 }
 
+extern TPWRAPPER_API unsigned getThorClusterNames(StringArray& targetNames, StringArray& queueNames)
+{
+#ifndef _CONTAINERIZED
+    StringArray thorNames, groupNames;
+    getEnvironmentThorClusterNames(thorNames, groupNames, targetNames, queueNames);
+#else
+    Owned<IStringIterator> targets = getContainerTargetClusters("thor", nullptr);
+    ForEach(*targets)
+    {
+        SCMStringBuffer target;
+        targets->str(target);
+        targetNames.append(target.str());
+
+        StringBuffer qName;
+        queueNames.append(getClusterThorQueueName(qName, target.str()));
+    }
+#endif
+    return targetNames.ordinality();
+}
+
+static BoolHash validTargets;
+static CriticalSection validTargetSect;
+
+extern TPWRAPPER_API void refreshValidTargets()
+{
+    validTargets.kill();
+#ifdef _CONTAINERIZED
+    // discovered from generated cluster names
+    Owned<IStringIterator> it = getContainerTargetClusters(nullptr, nullptr);
+#else
+    Owned<IStringIterator> it = getTargetClusters(nullptr, nullptr);
+#endif
+    ForEach(*it)
+    {
+        SCMStringBuffer s;
+        IStringVal& val = it->str(s);
+        bool* found = validTargets.getValue(val.str());
+        if (!found || !*found)
+        {
+            validTargets.setValue(val.str(), true);
+            PROGLOG("adding valid target: %s", val.str());
+        }
+    }
+}
+
+extern TPWRAPPER_API void validateTargetName(const char* target)
+{
+    if (isEmptyString(target))
+        throw makeStringException(ECLWATCH_INVALID_CLUSTER_NAME, "Empty target name.");
+
+    CriticalBlock block(validTargetSect);
+    bool* found = validTargets.getValue(target);
+    if (found && *found)
+        return;
+
+#ifndef _CONTAINERIZED
+    if (!validateTargetClusterName(target))
+#else
+    VStringBuffer xpath("queues[@name=\"%s\"]", target);
+    if (queryComponentConfig().queryPropTree(xpath) == nullptr)
+#endif
+        throw MakeStringException(ECLWATCH_INVALID_CLUSTER_NAME, "Invalid target name: %s", target);
+
+    refreshValidTargets();
```

Comment:
reloaded every time there's a mismatch? That sounds fairly wasteful.
In bare metal once loaded, does it ever need reloading? Only on an environment change? It could detect that..

In containerized mode, currently, if there's any change to the target queues, esp will be auto restarted by K8s.
In near future, we want to optimize that by spotting changes dynamically and refreshing internal structures/data like this. For now I'd just add a comment and never refresh.

Reply:
This is from the existing code. I do agree with you. I haven't seen an environment where the whole HPCC system does not need to restart after its environment setting changes (even for a bare metal environment). So, the lines 2271 - 2279 should be removed. Any objection?

Reply:
In cloud/containers - should load once and not refresh every mismatch at the moment - with a comment re. refreshing dynamically in future, but currently it will auto-restart the container.

In bare-metal, I'm not sure. What if the environment updates? Although wasteful, suspect should leave it as is, reloading on every mismatch in bare metal for now.